File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int kthSmallest(TreeNode root, int k) {
        Deque<Integer> s = new ArrayDeque<Integer>();
        s = getStack(root, s);
        int ans = -1;
        while(k > 0){
            ans = s.removeFirst();
            k--;
        }
        return ans;
    }

    public Deque<Integer> getStack(TreeNode root, Deque<Integer> s){
        if(root == null){
            return s;
        }

            getStack(root.left, s);
            s.addLast(root.val);
            getStack(root.right, s);


        return s;
    }


}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func goodNodes(root *TreeNode) int {
    return countGoodNodes(root.Val, root.Left) + countGoodNodes(root.Val, root.Right) +1
}

func countGoodNodes(t int, currNode *TreeNode) int {
	if currNode == nil {
		return 0
	}

	if currNode.Val < t {
		return countGoodNodes(t, currNode.Left) + countGoodNodes(t, currNode.Right)
	}
	t = maxInt(currNode.Val, t)
	return countGoodNodes(t, currNode.Left) + countGoodNodes(t, currNode.Right) + 1

}

func maxInt(a, b int) int{
	if a > b{
		return a
	}
	return b
}
